pkg/controller/predicate/rolloutgate.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicate
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	k8spredicate "sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// RolloutGate allows reconcilers to select which resources they are enabled for, based
+// on features of the resource to-be-reconciled. This allows for a gated rollout of the
+// controller.
+type RolloutGate interface {
+	// ShouldReconcile returns true if the reconciler should be used to for the resource.
+	ShouldReconcile(o *unstructured.Unstructured) bool
+}
+
+// RolloutPredicate generates a controller-runtime predicate based on a RolloutGate.
+type RolloutPredicate struct {
+	gvk schema.GroupVersionKind
+	rg  RolloutGate
+	c   client.Client
+}
+
+var _ k8spredicate.Predicate = &RolloutPredicate{}
+
+func NewRolloutPredicate(c client.Client, gvk schema.GroupVersionKind, rg RolloutGate) *RolloutPredicate {
+	return &RolloutPredicate{
+		c:   c,
+		gvk: gvk,
+		rg:  rg,
+	}
+}
+
+func (p *RolloutPredicate) Create(e event.CreateEvent) bool {
```

Comment:
I personally like to add a comment like `// Create implements the Predicate interface` just to make it clearer what we're doing.  I've seen it done elsewhere, and I assume at some stage there will be better support in IDEs for going to the interface method definition (so it's not worth doing more)...

Reply:
For me, I usually just add a "self-documenting" variable assignment (like here: https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/2559/files#diff-b3ba7e1f2676394700779aab805a0ac09590b4d96b002e9f84eb581915a6eb17R43) to show that it fits the interface: `var _ k8spredicate.Predicate = &RolloutPredicate{}`. Not sure if comments are really all that necessary